core/geometry/src/curve/internalContexts/geometry2d/ConstraintConstruction.ts

```diff
@@ -0,0 +1,271 @@
+    if (this._constraints.length !== 3)
+      return undefined;
+    const constraints = this.clone();
+    constraints.convertThroughPointToZeroRadiusCircles();
+    constraints.normalizeLines();
+    ConstraintSet.sortByConstraintTypeAndGeometryType(constraints._constraints);
+    // note that pass-through points are now tangentTo circles with zero radius
+    const numTangent = constraints.countConstraintType(ConstraintType.tangentTo);
+    const numRadius = constraints.countConstraintType(ConstraintType.radius);
+    const c0 = constraints._constraints[0];
+    const c1 = constraints._constraints[1];
+    const c2 = constraints._constraints[2];
+    if (numTangent === 3) {
+      // radius comes first, then line(s), then circle(s); once a circle is encountered, the rest must be circles
+      if (c0.curve instanceof UnboundedLine2dByPointAndNormal) {
+        if (c1.curve instanceof UnboundedLine2dByPointAndNormal) {
+          if (c2.curve instanceof UnboundedLine2dByPointAndNormal) {
+            return TangentConstruction.circlesTangentLLL(c0.curve, c1.curve, c2.curve);
+          } else if (c2.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+            return TangentConstruction.circlesTangentLLC(c0.curve, c1.curve, c2.curve);
+          }
+        } else if (c1.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+          if (c2.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+            return TangentConstruction.circlesTangentCCL(c1.curve, c2.curve, c0.curve);
+          }
+        }
+      } else if (c0.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+        if (c1.curve instanceof UnboundedCircle2dByCenterAndRadius
+          && c2.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+          return TangentConstruction.circlesTangentCCC(c0.curve, c1.curve, c2.curve);
+        }
+      }
+    } else if (numRadius === 1 && numTangent === 2 && c0.radius !== undefined) {
+      if (c1.curve instanceof UnboundedLine2dByPointAndNormal) {
+        if (c2.curve instanceof UnboundedLine2dByPointAndNormal) {
+          return TangentConstruction.circlesTangentLLR(c1.curve, c2.curve, c0.radius);
+        } else if (c2.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+          return TangentConstruction.circlesTangentCLR(c2.curve, c1.curve, c0.radius);
+        }
+      } else if (c1.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+        if (c2.curve instanceof UnboundedCircle2dByCenterAndRadius) {
+          return TangentConstruction.circlesTangentCCR(c1.curve, c2.curve, c0.radius);
+        }
+      }
+    }
+    return undefined;
+  }
+  /**
+   * Call constructConstrainedLines when the ConstraintSet has 2 conditions which determine one or more lines.
+   * @returns
+   */
+  public constructConstrainedLines(): ImplicitGeometryMarkup<UnboundedLine2dByPointAndNormal>[] | undefined {
```

Comment:
Please make sure to mention in the function doc that constraint set has to have 2 constraints.

Reply:
(pushed 8/26)